dependency-analysis/main.go

```diff
@@ -0,0 +1,349 @@
+}
+
+// GetDependencyDiff returns the dependency diff between two commits.
+// It returns an error if the dependency graph is not enabled.
+func GetDependencyDiff(owner, repo, token, base, head string) (*[]DependencyDiff, error) {
+	message := "failed to get dependency diff, please enable dependency graph https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/configuring-the-dependency-graph" //nolint:lll
+	if owner == "" {
+		return nil, fmt.Errorf("owner is required") //nolint:goerr113
+	}
+	if repo == "" {
+		return nil, fmt.Errorf("repo is required") //nolint:goerr113
+	}
+	if token == "" {
+		return nil, fmt.Errorf("token is required") //nolint:goerr113
+	}
+	resp, err := GetGitHubDependencyDiff(owner, repo, token, base, head)
+	defer resp.Body.Close() //nolint:staticcheck
+
+	if resp.StatusCode != http.StatusOK {
+		// if the dependency graph is not enabled, we can't get the dependency diff
+		return nil,
+			fmt.Errorf(" %s: %v", message, resp.Status) //nolint:goerr113
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to get dependency diff: %w", err)
+	}
+
+	var data []DependencyDiff
+	err = json.NewDecoder(resp.Body).Decode(&data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode dependency diff: %w", err)
+	}
+	// filter out the dependencies that are not added
+	var filteredData []DependencyDiff
+	for _, dep := range data { //nolint:gocritic
+		// also if the source repo doesn't start with GitHub.com, we can ignore it
+		if dep.ChangeType == "added" && dep.SourceRepositoryURL != "" &&
+			strings.HasPrefix(dep.SourceRepositoryURL, "https://github.com") {
+			filteredData = append(filteredData, dep)
+		}
+	}
+	return &filteredData, nil
+}
+
+// GetGitHubDependencyDiff returns the dependency diff between two commits.
+// It returns an error if the dependency graph is not enabled.
+func GetGitHubDependencyDiff(owner, repo, token, base, head string) (*http.Response, error) {
+	req, err := http.NewRequest("GET",
+		fmt.Sprintf("https://api.github.com/repos/%s/%s/dependency-graph/compare/%s...%s", owner, repo, base, head), nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %w", err)
+		// handle err
+	}
+	req.Header.Set("Accept", "application/vnd.github+json")
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get response: %w", err)
+	}
+	return resp, nil
+}
+
+// filter returns a new slice containing all elements of slice that satisfy the predicate f.
+func filter[T any](slice []T, f func(T) bool) []T {
+	var n []T
+	for _, e := range slice {
+		if f(e) {
+			n = append(n, e)
+		}
+	}
+	return n
+}
+
+// GetScorecardChecks returns the list of checks to run.
+// This uses the SCORECARD_CHECKS environment variable to get the path to the checks list.
+func GetScorecardChecks() ([]string, error) {
+	fileName := os.Getenv("SCORECARD_CHECKS")
```

Comment:
Agree that it should be parameterized for testability; also wonder if filename is the right approach? If the user is configuring the DependencyDiff action, can their string environment variable be the comma-separated list of checks instead of the name of a file with the list of checks?

Reply:
> would it make sense to move filename to a paremeter?
> That would centralize all the environment variable reading in `main()` and would be easier to test this function.

I can move the ENV variable to main. 
>also wonder if filename is the right approach? If the user is configuring the DependencyDiff action, can their string environment variable be the comma-separated list of checks instead of the name of a file with the list of checks?

The reason for the file is to make is cleaner to configure instead of passing `19` checks as an argument in the command line. 

Also when we have the policy it is also going to be a file and that was my thought process